File: backend/src/lib/graph-topology.ts
import type { DecisionEvent } from '../schemas/decision.schema';

export type AgentRole = 'publisher' | 'seeker' | 'main';

export function userSlugFromNodeId(nodeId: string): string {
  const m = nodeId.match(/^User_(.+)$/i);
  if (m) return `user_${m[1].toLowerCase()}`;
  return nodeId.toLowerCase().replace(/[^a-z0-9]+/g, '_');
}

export function userNodeIdFromSlug(slug: string): string {
  /** Gemini / slug-style ids (e.g. user_117) stay as-is. */
  if (/^user_[a-z0-9_]+$/i.test(slug) && slug === slug.toLowerCase()) return slug;
  if (slug.startsWith('user_')) return `User_${slug.slice(5)}`;
  return slug;
}

export function agentNodeId(userSlug: string, role: AgentRole): string {
  return `${userSlug}.${role}`;
}

export function graphIdFor(userSlug: string, role: AgentRole, version = 'v1'): string {
  return `${userSlug}.${role}.${version}`;
}

export function parseGraphId(
  graphId: string,
): { userSlug: string; role: AgentRole; version: string } | null {
  const m = graphId.match(/^(.+)\.(publisher|seeker|main)\.(v\d+)$/);
  if (!m) return null;
  return { userSlug: m[1], role: m[2] as AgentRole, version: m[3] };
}

export function primaryUserNodeId(nodes: { node_type: string; node_id: string }[]): string | null {
  const user = nodes.find((n) => n.node_type === 'user');
  return user?.node_id ?? null;
}

export function resolveAgentContext(event: DecisionEvent): {
  userSlug: string;
  role: AgentRole;
  agentId: string;
  userNodeId: string;
} | null {
  const parsed = parseGraphId(event.graph_id);
  if (parsed) {
    return {
      userSlug: parsed.userSlug,
      role: parsed.role,
      agentId: agentNodeId(parsed.userSlug, parsed.role),
      userNodeId: userNodeIdFromSlug(parsed.userSlug),
    };
  }
  const userNodeId = primaryUserNodeId(event.nodes);
  if (!userNodeId) return null;
  const userSlug = userSlugFromNodeId(userNodeId);
  return {
    userSlug,
    role: 'publisher',
    agentId: agentNodeId(userSlug, 'publisher'),
    userNodeId,
  };
}

File: backend/app/lib/graph_topology.py
import re
from typing import Literal, Optional

from app.schemas.decision import DecisionEvent

AgentRole = Literal["publisher", "seeker", "main"]


def user_slug_from_node_id(node_id: str) -> str:
    m = re.match(r"^User_(.+)$", node_id, re.I)
    if m:
        return f"user_{m.group(1).lower()}"
    return re.sub(r"[^a-z0-9]+", "_", node_id.lower())


def user_node_id_from_slug(slug: str) -> str:
    if re.match(r"^user_[a-z0-9_]+$", slug, re.I) and slug == slug.lower():
        return slug
    if slug.startswith("user_"):
        return f"User_{slug[5:]}"
    return slug


def agent_node_id(user_slug: str, role: AgentRole) -> str:
    return f"{user_slug}.{role}"


def graph_id_for(user_slug: str, role: AgentRole, version: str = "v1") -> str:
    return f"{user_slug}.{role}.{version}"


def parse_graph_id(graph_id: str) -> Optional[dict[str, str]]:
    m = re.match(r"^(.+)\.(publisher|seeker|main)\.(v\d+)$", graph_id)
    if not m:
        return None
    return {"userSlug": m.group(1), "role": m.group(2), "version": m.group(3)}


def primary_user_node_id(nodes: list) -> Optional[str]:
    for n in nodes:
        if n.node_type == "user":
            return n.node_id
    return None


def resolve_agent_context(event: DecisionEvent) -> Optional[dict]:
    parsed = parse_graph_id(event.graph_id)
    if parsed:
        return {
            "userSlug": parsed["userSlug"],
            "role": parsed["role"],
            "agentId": agent_node_id(parsed["userSlug"], parsed["role"]),
            "userNodeId": user_node_id_from_slug(parsed["userSlug"]),
        }
    user_node_id = primary_user_node_id(event.nodes)
    if not user_node_id:
        return None
    user_slug = user_slug_from_node_id(user_node_id)
    return {
        "userSlug": user_slug,
        "role": "publisher",
        "agentId": agent_node_id(user_slug, "publisher"),
        "userNodeId": user_node_id,
    }
